### PR TITLE
[GridLayer] Add getElevationValue to enable grid elevation by aggregation

### PR DESCRIPTION
**Pull request by heshan0131**

### CODE REVIEW

File: src/core-layers/grid-layer/grid-layer.js [hunk 1]
```diff
@@ -56,114 +69,238 @@ export default class GridLayer extends CompositeLayer {
...
+    } else if (dimensionChanges) {
+      dimensionChanges.forEach(f => typeof f === 'function' && f.apply(this));
+    }
+  }
 
-      // this needs sortedBins to be set
-      this.getValueDomain();
-    } else if (this.needsReSortBins(oldProps, props)) {
+  needsReProjectPoints(oldProps, props) {
+    return oldProps.cellSize !== props.cellSize || oldProps.getPosition !== props.getPosition;
```
Comment (ibgreen) [inline src/core-layers/grid-layer/grid-layer.js]: This line is probably my only big concern with this PR. This goes against how deck.gl layers usually work. Supplying a new accessor should not trigger an update. We provide fine grained control over that with `updateTriggers`.

I have written a new article about [updates](https://github.com/uber/deck.gl/blob/master/docs/advanced/updates.md) for the upcoming release to make this more clear to users.

File: src/core-layers/grid-layer/grid-layer.js [hunk 2]
```diff
@@ -56,114 +69,238 @@ export default class GridLayer extends CompositeLayer {
...
+      object
+    });
   }
 
-  needsReSortBins(oldProps, props) {
-    return oldProps.getColorValue !== props.getColorValue;
+  getUpdateTriggers() {
+    const {dimensionUpdaters} = this.state;
+    // merge all dimension triggers
+    return Object.keys(dimensionUpdaters).reduce((accu, dimension) => {
```
Comment (ibgreen) [inline src/core-layers/grid-layer/grid-layer.js]: Nit: This is basically just opinions, so take them with a grain of salt:

I find this code is hard to follow - partly because I am personally not a fan of `reduce` calls, in spite of their theoretical appeal I find them hard to read, especially the JavaScript `reduce` syntax.

But more importantly, even if this was a `for of loop`, again it feels like this type of generic iteration code would belong in a generic aggregation manager, where it would make perfect sense, but in a layer I would rather see code that explicitly treated each attribute, and kept things concrete. 

Put another way: Are we really saving a lot of typing by raising the abstraction level like this?

**Comment by akre54** [comment]
:thumbsup: on a generic AggregationManager. This would set the stage for extensibility on things like knn clusters, heatmaps, or grids. I don't think it's a blocker for this release but definitely important for future releases.

File: src/core-layers/grid-layer/grid-layer.js [hunk 3]
```diff
@@ -56,114 +69,238 @@ export default class GridLayer extends CompositeLayer {
...
 
-      this.getSortedBins();
-      this.getValueDomain();
+  getDimensionUpdaters() {
+    // dimension updaters are sequential,
+    // if the first one needs to be called, the 2nd and 3rd one will automatically
+    // be called. e.g. if ColorValue needs to be updated, getColorValueDomain and getColorScale
+    // will automatically be called
+    return {
+      getColor: [
```
[the remark this answers is not in the document]
Comment (heshan0131) [reply src/core-layers/grid-layer/grid-layer.js]: will setup a meeting

File: docs/layers/grid-layer.md [hunk 4]
```diff
@@ -113,6 +113,19 @@ Elevation scale input domain, default is set to the extent of point counts in ea
...
 
+##### `getElevationValue` (Function, optional)
+
+- Default: `points => points.length`
+
+Similar to `getColorValue`, `getElevationValue` is the accessor function to get the value that cell elevation is based on.
+It takes an array of points inside each cell as arguments, returns a number.
+By default `getElevationValue` returns the length of the points array.
+
+Note: grid layer compares whether `getElevationValue` has changed to
```
[the remark this answers is not in the document]
Comment (heshan0131) [reply docs/layers/grid-layer.md]: there is lot of effort made to optimize aggregation performance, and to see it in action, I feel like it is necessary to mention here this accessor is always better to be defined outside. Moving forward, I can make this also controlled by updateTriggers (hence aggregation manager), but I feel like for this type of accessor, at least for these 2 layers, it's necessary to mention the performance improvement here

File: docs/layers/grid-layer.md [hunk 5]
```diff
@@ -113,6 +113,19 @@ Elevation scale input domain, default is set to the extent of point counts in ea
 
 Elevation scale output range
 
+##### `getElevationValue` (Function, optional)
```
[the remark this answers is not in the document]
Comment (heshan0131) [reply docs/layers/grid-layer.md]: done

File: docs/layers/grid-layer.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (ibgreen) [reply docs/layers/grid-layer.md]: Sounds good. Maybe add that context to the docs
```
Aggregation is particularly performance sensitive, so you might want to consider...
```

File: src/core-layers/grid-layer/grid-layer.js (hunk 1 above)
Comment (heshan0131) [reply src/core-layers/grid-layer/grid-layer.js]: will remove this and rely on user supply updateTriggers
